src/pevm.rs

```diff
@@ -263,10 +266,16 @@ impl Pevm {
                         _ => unreachable!(),
                     }
                     // Assert that evaluated nonce is correct when address is caller.
-                    debug_assert!(
-                        tx.caller != address || tx.nonce.map_or(true, |n| n + 1 == info.nonce)
-                    );
-
+                    if tx.caller == address {
+                        if let Some(nonce) = tx.nonce {
+                            if nonce + 1 != info.nonce {
+                                return Err(PevmError::NonceMismatch {
```

Comment:
Let's add a `// TODO: Consider falling back to sequential instead`

Reply:
If `PevmError::NonceMismatch` happens, it is likely that the input is invalid (i.e. because the block building service is feeding pevm with a nonce-too-low or nonce-too-high tx).

So I think we should not "consider falling back to sequential". We should `return Err(...)` instead.

Reply:
No, block building shouldn't wait until post-processing to arrive at an error that needs to rework completely. Nonce checking should be done before or during block building instead. On the other hand, nonce errors during post-processing is possible if there's a bug in lazy evaluation or some weird race conditions that we haven't addressed.

Reply:
comment added